zerocopy-derive/src/enums.rs

```diff
@@ -0,0 +1,491 @@
+
+    let discriminant_ident = &idents.discriminant;
+
+    let match_arms = data.variants.iter().map(|variant| {
+        let value_ident = idents.discriminant_const(&variant.ident);
+
+        if matches!(variant.fields, Fields::Unit) {
+            quote! {
+                #value_ident => true
+            }
+        } else {
+            let variant_struct = idents.variant_struct(&variant.ident);
+
+            quote! {
+                #value_ident => {
+                    // SAFETY:
+                    // - `cast_unsized`:
+                    //   - The variant struct has the same size as the enum
+                    //   - The returned pointer is cast from `p`, and so has the
+                    //     same provenance as `p`.
+                    //   - We checked that the discriminant of the enum matched
+                    //     the value for this variant, so this cast preserves
+                    //     types and locations of all fields. Therefore, any
+                    //     `UnsafeCell`s will have the same location as in the
+                    //     original type.
+                    // - `assume_initialized`: `cast_unsized` removes the
+                    //   initialization invariant from `p`, so we re-assert that
+                    //   all of the bytes are initialized.
+                    let candidate = unsafe {
+                        candidate
+                            .cast_unsized(|p: *mut Self| p as *mut #variant_struct)
+                            .assume_initialized()
+                    };
+                    #variant_struct::is_bit_valid(candidate)
+                }
+            }
+        }
+    });
+
+    quote! {
+        // SAFETY: We use `is_bit_valid` to validate that the bit pattern
+        // corresponds to one of the enum's variant discriminants. Then, we
+        // check the bit validity of each field of that variant. Thus, this is a
+        // sound implementation of `is_bit_valid`.
+        fn is_bit_valid<A>(
+            mut candidate: ::zerocopy::Maybe<'_, Self, A>,
+        ) -> ::zerocopy::macro_util::core_reexport::primitive::bool
+        where
+            A: ::zerocopy::pointer::invariant::Aliasing
+                + ::zerocopy::pointer::invariant::AtLeast<::zerocopy::pointer::invariant::Shared>,
+        {
+            use ::zerocopy::macro_util::core_reexport;
+
+            let candidate = candidate.reborrow().forget_exclusive();
+
+            #discriminant_type
+            #discriminant_consts
+            #(#variant_structs)*
+
+            // SAFETY:
+            // - `cast` is implemented as required.
+            // - By definition, `*mut Self` and `*mut [u8; size_of::<Self>()]`
+            //   are types of the same size.
+            let discriminant = unsafe { candidate.cast_unsized(|p: *mut Self| p as *mut [core_reexport::primitive::u8; core_reexport::mem::size_of::<#discriminant_ident>()]) };
```

Comment:
For `#[repr(P)]`, could we read this as a `P` instead of a `u8` array? It would maybe allow us to generate better code since we can assume alignment.

Reply:
Are the sizes of `#[repr(C)]` enums well-defined?

Reply:
Yeah, it'd need to be just for `repr(u8)`, `repr(i16)`, etc.

Reply:
I checked the implementation in rustc and it says that `#[repr(C)]` enums always have an `isize` discriminant. That's not documented AFAICT, but I can insert a const-eval assertion that `size_of::<EnumTag>() == size_of::<isize>()` so we fail gracefully if it's not `isize`-sized. Does that sound good?

Reply:
That'd be a stability hazard - if they ever change it, older versions of zerocopy would break. I'd vote that we remain conservative for `repr(C)`, but we should also put up a PR to the reference to document that as a guarantee, and then we can do what you proposed once it lands.

Reply:
After reading more closely, the integer primitive for `repr(C)` enum discriminants is Very Complicated. It's at least as large as `c_int::BITS` but bigger if it needs to be, and unsigned if possible but can be signed if necessary. So we can't correctly guess the type of the discriminant in most cases.

I think it would be very disappointing if we had worse codegen for `repr(C)` than for other enums. I think `repr(C)` enums will be pretty common, but I could be wrong. We could work around this with something like:

```rust
struct Int;

trait OfSize<const N: usize> {
    type Type;
}

impl OfSize<1> for Int {
    type Type = i8;
}

impl OfSize<2> for Int {
    type Type = i16;
}

impl OfSize<4> for Int {
    type Type = i32;
}

impl OfSize<8> for Int {
    type Type = i64;
}

impl OfSize<16> for Int {
    type Type = i128;
}

// in macro:

#[repr(C)]
enum Discriminant {
    A,
    // ...
}

const SIZE: usize = ::core::mem::size_of::<Discriminant>();
type DiscInt = <Int as OfSize<SIZE>>::Type;

const A_DISC: DiscInt = DiscInt::from_ne_bytes(
    unsafe { transmute::<Foo, [u8; SIZE]>(Foo::A) }
);
// ...
```
And then we can read the discriminant as `DiscInt` and match against the discriminant consts. We can get away with this because the generated discriminant enum doesn't have any generics. Does that sound more appealing?

Reply:
Didn't realize it'd be this complicated. Let's just keep it as a u8 array in this PR, but file an issue to follow up with an improvement as described here in a future PR?